# DATA ANALYSIS PROJECT

# 🛒 AVIATION SECTOR Data Analysis Project

## 📌 Project Objective

This project demonstrates an end-to-end retail data analysis workflow, beginning with raw transactional data and progressing through data auditing, cleaning, validation, exploratory data analysis (EDA), and business insights.

## Goals

- Understand the dataset
- Identify data quality issues
- Handle missing values
- Validate business rules
- Perform exploratory data analysis
- Generate business recommendations

## Tools Used
- Python
- Pandas
- NumPy
- Matplotlib
- Seaborn

# 📑 Notebook Roadmap

## Import Libraries
- Load Dataset
- Dataset Overview
- Data Description
- Data Audit
- Missing Value Analysis
- Data Cleaning
- Data Validation
- Feature Engineering
- Exploratory Data Analysis
- Business Insights
- Recommendations
- Conclusion

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
df=pd.read_csv('data1_raw.csv',low_memory=False)
df.head()

,Year,Quarter,Month,DayofMonth,DayOfWeek,FlightDate,Reporting_Airline,DOT_ID_Reporting_Airline,IATA_CODE_Reporting_Airline,Tail_Number,...,Div4WheelsOff,Div4TailNum,Div5Airport,Div5AirportID,Div5AirportSeqID,Div5WheelsOn,Div5TotalGTime,Div5LongestGTime,Div5WheelsOff,Div5TailNum
0,2025,1,1,1,3,01-01-2025,AA,19805,AA,N104NN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2025,1,1,2,4,02-01-2025,AA,19805,AA,N110AN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,2025,1,1,3,5,03-01-2025,AA,19805,AA,N106NN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,2025,1,1,4,6,04-01-2025,AA,19805,AA,N117AN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,2025,1,1,5,7,05-01-2025,AA,19805,AA,N104NN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


## grain/granularity of the dataset
-- what one row represent

In [3]:
# SHAPE
df.shape


(539747, 109)

In [4]:
# Basic info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 539747 entries, 0 to 539746
Columns: 109 entries, Year to Div5TailNum
dtypes: float64(64), int64(26), object(19)
memory usage: 448.9+ MB


In [5]:
# list of all cloumns as there are 109 
df.columns.tolist()

['Year',
 'Quarter',
 'Month',
 'DayofMonth',
 'DayOfWeek',
 'FlightDate',
 'Reporting_Airline',
 'DOT_ID_Reporting_Airline',
 'IATA_CODE_Reporting_Airline',
 'Tail_Number',
 'Flight_Number_Reporting_Airline',
 'OriginAirportID',
 'OriginAirportSeqID',
 'OriginCityMarketID',
 'Origin',
 'OriginCityName',
 'OriginState',
 'OriginStateFips',
 'OriginStateName',
 'OriginWac',
 'DestAirportID',
 'DestAirportSeqID',
 'DestCityMarketID',
 'Dest',
 'DestCityName',
 'DestState',
 'DestStateFips',
 'DestStateName',
 'DestWac',
 'CRSDepTime',
 'DepTime',
 'DepDelay',
 'DepDelayMinutes',
 'DepDel15',
 'DepartureDelayGroups',
 'DepTimeBlk',
 'TaxiOut',
 'WheelsOff',
 'WheelsOn',
 'TaxiIn',
 'CRSArrTime',
 'ArrTime',
 'ArrDelay',
 'ArrDelayMinutes',
 'ArrDel15',
 'ArrivalDelayGroups',
 'ArrTimeBlk',
 'Cancelled',
 'CancellationCode',
 'Diverted',
 'CRSElapsedTime',
 'ActualElapsedTime',
 'AirTime',
 'Flights',
 'Distance',
 'DistanceGroup',
 'CarrierDelay',
 'WeatherDelay',
 'NASDelay',
 'SecurityD

# WORKING ASSUMPTIONS
-- One row means one flight operation as an assumption

In [6]:
df[['FlightDate', 'Reporting_Airline', 'Flight_Number_Reporting_Airline',
    'Origin', 'Dest']].head(10)

,FlightDate,Reporting_Airline,Flight_Number_Reporting_Airline,Origin,Dest
0,01-01-2025,AA,1,JFK,LAX
1,02-01-2025,AA,1,JFK,LAX
2,03-01-2025,AA,1,JFK,LAX
3,04-01-2025,AA,1,JFK,LAX
4,05-01-2025,AA,1,JFK,LAX
5,06-01-2025,AA,1,JFK,LAX
6,07-01-2025,AA,1,JFK,LAX
7,08-01-2025,AA,1,JFK,LAX
8,09-01-2025,AA,1,JFK,LAX
9,10-01-2025,AA,1,JFK,LAX


In [7]:
df[['FlightDate', 'Reporting_Airline', 'Flight_Number_Reporting_Airline',
    'Origin', 'Dest']].duplicated().sum()

np.int64(0)

### INSIGHT
SINCE THERE IS NO DUPLICATE VALUES FOR MY ASSUMPTION THEREFORE MY ASSUMPTION IS CORRECT AND THUS 1 row = 1 flight on a particular date, operated by an airline, from an origin to a destination.

In [8]:
# DUPLICATE FLIGHT NUMBER REPORT
df['Flight_Number_Reporting_Airline'].isna().sum()

np.int64(0)

In [9]:
df['Flight_Number_Reporting_Airline'].nunique()

6390

# CONCLUSION
One row represents one flight operation identified by flight date, reporting airline, flight number, origin airport, and destination airport.

### COLUMN AUDIT -- REMOVING COLUMNS THAT ARE NOT NEEDED

In [10]:
df.isna().sum().sort_values(ascending=False) # THERE ARE SOME COLUMNS WITH COMPLETE NULL DATA

Div5TailNum         539747
Div4TotalGTime      539747
Div3Airport         539747
Div3AirportID       539747
Div3AirportSeqID    539747
                     ...  
DestStateFips            0
DestState                0
DestCityName             0
Dest                     0
Distance                 0
Length: 109, dtype: int64

In [11]:
missing_df = (
    df.isna()
      .mean()
      .mul(100)
      .sort_values(ascending=False)
      .reset_index()
)

missing_df.columns = ['column', 'missing_pct']

missing_df

,column,missing_pct
0,Div5TailNum,100.0
1,Div4TotalGTime,100.0
2,Div3Airport,100.0
3,Div3AirportID,100.0
4,Div3AirportSeqID,100.0
...,...,...
104,DestStateFips,0.0
105,DestState,0.0
106,DestCityName,0.0
107,Dest,0.0


In [12]:
# COMPLETE MISSING
missing_df[missing_df['missing_pct']==100].head()

,column,missing_pct
0,Div5TailNum,100.0
1,Div4TotalGTime,100.0
2,Div3Airport,100.0
3,Div3AirportID,100.0
4,Div3AirportSeqID,100.0


In [13]:
### THERE ARE TOTAL 24 COLUMNS WITH COMPLETE NULL DATA
(df.isna().mean() == 1).sum()

np.int64(24)

In [14]:
# REMOVING COLUMNS THAT ARE 100 % EMPTY
empty_cols = df.columns[df.isna().all()].tolist()
len(empty_cols)
# df=df.drop(columns=empty_cols)

24

In [15]:
df = df.drop(columns=empty_cols)

In [16]:
df.shape

(539747, 85)

In [17]:
(df.isna().mean() == 1).sum()

np.int64(0)

### CLEANED COLUMNS WITH 100 % MISSING VALUES 

# CLEANING REMAINING COLUMNS

In [18]:
# professional data missing report 
from IPython.display import display
import pandas as pd
missing_report=pd.DataFrame(
    {
        'Missing Count': df.isna().sum(),
    'Missing %': df.isna().mean() * 100,
    'Data Type': df.dtypes
    }
    
)
missing_report = (
    missing_report
    .sort_values('Missing %', ascending=False)
)

display(missing_report)

,Missing Count,Missing %,Data Type
Div2TailNum,539743,99.999259,object
Div2WheelsOff,539743,99.999259,float64
Div2LongestGTime,539724,99.995739,float64
Div2TotalGTime,539724,99.995739,float64
Div2WheelsOn,539724,99.995739,float64
...,...,...,...
DestStateFips,0,0.000000,int64
DestState,0,0.000000,object
Dest,0,0.000000,object
DestCityMarketID,0,0.000000,int64


In [19]:
# PROBLEM 100 % MISSING DATA COLUMNS SHOULD BE DELETED BUT STILL SHOWING b
missing_report = (
    pd.DataFrame({
        'Column': df.columns,
        'Missing Count': df.isna().sum().values,
        'Missing %': df.isna().mean().mul(100).values,
        'Data Type': df.dtypes.astype(str).values
    })
    .sort_values('Missing %', ascending=False)
    .reset_index(drop=True)
)

missing_report['Missing %'] = missing_report['Missing %'].round(2)

display(missing_report)

,Column,Missing Count,Missing %,Data Type
0,Div2TailNum,539743,100.0,object
1,Div2WheelsOff,539743,100.0,float64
2,Div2LongestGTime,539724,100.0,float64
3,Div2TotalGTime,539724,100.0,float64
4,Div2WheelsOn,539724,100.0,float64
...,...,...,...,...
80,DestStateFips,0,0.0,int64
81,DestState,0,0.0,object
82,Dest,0,0.0,object
83,DestCityMarketID,0,0.0,int64


In [20]:
# COLUMNS WITH EMPTY DATA
missing_only = missing_report[
    missing_report['Missing Count'] > 0
]

display(missing_only)

,Column,Missing Count,Missing %,Data Type
0,Div2TailNum,539743,100.00,object
1,Div2WheelsOff,539743,100.00,float64
2,Div2LongestGTime,539724,100.00,float64
3,Div2TotalGTime,539724,100.00,float64
4,Div2WheelsOn,539724,100.00,float64
5,Div2AirportSeqID,539724,100.00,float64
6,Div2AirportID,539724,100.00,float64
7,Div2Airport,539724,100.00,object
8,DivArrDelay,538849,99.83,float64
9,DivActualElapsedTime,538849,99.83,float64


In [21]:
missing_report = (
    pd.DataFrame({
        'Column': df.columns,
        'Missing Count': df.isna().sum().values,
        'Missing %': df.isna().mean().mul(100).values,
        'Data Type': df.dtypes.astype(str).values
    })
    .sort_values('Missing %', ascending=False)
    .reset_index(drop=True)
)

missing_report['Missing %'] = missing_report['Missing %'].round(2)

display(missing_report)

,Column,Missing Count,Missing %,Data Type
0,Div2TailNum,539743,100.0,object
1,Div2WheelsOff,539743,100.0,float64
2,Div2LongestGTime,539724,100.0,float64
3,Div2TotalGTime,539724,100.0,float64
4,Div2WheelsOn,539724,100.0,float64
...,...,...,...,...
80,DestStateFips,0,0.0,int64
81,DestState,0,0.0,object
82,Dest,0,0.0,object
83,DestCityMarketID,0,0.0,int64


In [22]:
display(
    missing_report[missing_report['Missing %'] == 100]
)

,Column,Missing Count,Missing %,Data Type
0,Div2TailNum,539743,100.0,object
1,Div2WheelsOff,539743,100.0,float64
2,Div2LongestGTime,539724,100.0,float64
3,Div2TotalGTime,539724,100.0,float64
4,Div2WheelsOn,539724,100.0,float64
5,Div2AirportSeqID,539724,100.0,float64
6,Div2AirportID,539724,100.0,float64
7,Div2Airport,539724,100.0,object


In [23]:

missing_report = (
    pd.DataFrame({
        'Column': df.columns,
        'Missing Count': df.isna().sum().values,
        'Missing %': df.isna().mean().mul(100).values,
        'Data Type': df.dtypes.astype(str).values
    })
    .sort_values('Missing %', ascending=False)
    .reset_index(drop=True)
)

missing_report['Missing %'] = missing_report['Missing %'].round(2)

display(missing_report.head(15))

,Column,Missing Count,Missing %,Data Type
0,Div2TailNum,539743,100.00,object
1,Div2WheelsOff,539743,100.00,float64
2,Div2LongestGTime,539724,100.00,float64
3,Div2TotalGTime,539724,100.00,float64
4,Div2WheelsOn,539724,100.00,float64
5,Div2AirportSeqID,539724,100.00,float64
6,Div2AirportID,539724,100.00,float64
7,Div2Airport,539724,100.00,object
8,DivArrDelay,538849,99.83,float64
9,DivActualElapsedTime,538849,99.83,float64


In [24]:
print("CURRENT DF")
print("Shape:", df.shape)
print("100% empty columns:", df.columns[df.isna().all()].tolist())

print("\nCURRENT REPORT")
print("Shape:", missing_report.shape)
print("Columns in report:", missing_report['Column'].tolist()[:15])

CURRENT DF
Shape: (539747, 85)
100% empty columns: []

CURRENT REPORT
Shape: (85, 4)
Columns in report: ['Div2TailNum', 'Div2WheelsOff', 'Div2LongestGTime', 'Div2TotalGTime', 'Div2WheelsOn', 'Div2AirportSeqID', 'Div2AirportID', 'Div2Airport', 'DivArrDelay', 'DivActualElapsedTime', 'Div1TailNum', 'Div1WheelsOff', 'DivReachedDest', 'DivDistance', 'Div1AirportSeqID']


### REASON -- THERE IS A MARGIN OF 0.0009 approx to 100 thus it was a problem


In [25]:
# CATEGORIZING MISSING VALUE COLUMN
missing_summary=pd.DataFrame(
    {
        'Category':[
              'No Missing Values',
        'Less than 5% Missing',
        '5%–50% Missing',
        '50%–95% Missing',
        'More than 95% Missing'
        ],
        'Column Count':[
            (df.isna().mean() == 0).sum(),
        ((df.isna().mean() > 0) & (df.isna().mean() < 0.05)).sum(),
        ((df.isna().mean() >= 0.05) & (df.isna().mean() < 0.50)).sum(),
        ((df.isna().mean() >= 0.50) & (df.isna().mean() < 0.95)).sum(),
        (df.isna().mean() >= 0.95).sum()
        ]
    }
        
)
display(missing_summary)

,Category,Column Count
0,No Missing Values,39
1,Less than 5% Missing,17
2,5%–50% Missing,0
3,50%–95% Missing,5
4,More than 95% Missing,24


## FIXING CATEGORY -- MORE THAN 95 % MISSING

In [26]:
high_missing=(
    missing_report[missing_report['Missing %']>95][['Column', 'Missing Count', 'Missing %', 'Data Type']].copy()
) 
display(high_missing)

,Column,Missing Count,Missing %,Data Type
0,Div2TailNum,539743,100.00,object
1,Div2WheelsOff,539743,100.00,float64
2,Div2LongestGTime,539724,100.00,float64
3,Div2TotalGTime,539724,100.00,float64
4,Div2WheelsOn,539724,100.00,float64
5,Div2AirportSeqID,539724,100.00,float64
6,Div2AirportID,539724,100.00,float64
7,Div2Airport,539724,100.00,object
8,DivArrDelay,538849,99.83,float64
9,DivActualElapsedTime,538849,99.83,float64


In [27]:
# SINCE MOSTLY DIRVERSION RELATED DATA
diversion_profile = pd.DataFrame({
    'Column': [
        'DivReachedDest',
        'DivDistance',
        'DivActualElapsedTime',
        'DivArrDelay',
        'Div1Airport',
        'Div1TailNum',
        'Div2Airport'
    ],
    'Non-Missing Count': [
        df['DivReachedDest'].notna().sum(),
        df['DivDistance'].notna().sum(),
        df['DivActualElapsedTime'].notna().sum(),
        df['DivArrDelay'].notna().sum(),
        df['Div1Airport'].notna().sum(),
        df['Div1TailNum'].notna().sum(),
        df['Div2Airport'].notna().sum()
    ]
})

diversion_profile['Non-Missing %'] = (
    diversion_profile['Non-Missing Count'] / len(df) * 100
).round(4)

display(diversion_profile)

,Column,Non-Missing Count,Non-Missing %
0,DivReachedDest,1166,0.2160
1,DivDistance,1166,0.2160
2,DivActualElapsedTime,898,0.1664
3,DivArrDelay,898,0.1664
4,Div1Airport,1252,0.2320
5,Div1TailNum,918,0.1701
6,Div2Airport,23,0.0043


In [28]:
diversion_drop_cols = [
    'Div2TailNum',
    'Div2WheelsOff',
    'Div2LongestGTime',
    'Div2TotalGTime',
    'Div2WheelsOn',
    'Div2AirportSeqID',
    'Div2AirportID',
    'Div2Airport',
    
    'Div1AirportSeqID',
    'Div1AirportID',
    'Div1WheelsOn',
    'Div1WheelsOff',
    'Div1TotalGTime',
    'Div1LongestGTime'
]

display(
    pd.DataFrame({
        'Column to Drop': diversion_drop_cols
    })
)

print(f"Columns proposed for removal: {len(diversion_drop_cols)}")

,Column to Drop
0,Div2TailNum
1,Div2WheelsOff
2,Div2LongestGTime
3,Div2TotalGTime
4,Div2WheelsOn
5,Div2AirportSeqID
6,Div2AirportID
7,Div2Airport
8,Div1AirportSeqID
9,Div1AirportID


Columns proposed for removal: 14


In [29]:
# REMOVING THESE COLUMNS AS THEY HAVE VERY LESS DEPENDENCY
df = df.drop(columns=diversion_drop_cols)


In [30]:
df.shape

(539747, 71)

In [31]:
# CLEANED
cleanup_summary = pd.DataFrame({
    'Metric': [
        'Columns before',
        'Columns removed',
        'Columns after',
        'Rows'
    ],
    'Value': [
        85,
        len(diversion_drop_cols),
        df.shape[1],
        df.shape[0]
    ]
})

display(cleanup_summary)

,Metric,Value
0,Columns before,85
1,Columns removed,14
2,Columns after,71
3,Rows,539747


### NON DIVERSION RELATED COLUMNS -- CancellationCode, FirstDepTime, TotalAddGTime

In [32]:
# FOR CANCELLATION CODE -- MISSING DUE TO ERROR OR THERE WAS FARSE DATA
cancellation_analysis=(
    df.groupby('Cancelled')['CancellationCode'].agg(
        Total_Flights='size',
          Missing_Codes=lambda x: x.isna().sum(),
          Non_Missing_Codes=lambda x: x.notna().sum()
    ).reset_index()
)
# IN PERCENTAGE
cancellation_analysis['Missing %'] = (
    cancellation_analysis['Missing_Codes']
    / cancellation_analysis['Total_Flights']
    * 100
).round(2)

display(cancellation_analysis)

,Cancelled,Total_Flights,Missing_Codes,Non_Missing_Codes,Missing %
0,0,523435,523435,0,100.0
1,1,16312,0,16312,0.0


In [33]:
## HYPOSTHESIS IF VALUE == 1 THERE SHOULD BE A CANCELLATION CODE 
                # IF VALUE ==0 THERE SHOULD BE NO CODE

### INSIGHTS
- CancellationCode = NaN isn't necessarily "missing information." IT REPRESENTS NOT CANCELLED
- 523,435 / 523,435 non-cancelled flights → missing
16,312 / 16,312 cancelled flights → populated

In [34]:
# DELAY REASON COLUMNS -- CAUSING VALUES TO BE MISSING FROM CODE
delay_reason_cols = [
    'CarrierDelay',
    'WeatherDelay',
    'NASDelay',
    'SecurityDelay',
    'LateAircraftDelay'
]
delay_analysis = (
    df[delay_reason_cols]
    .notna()
    .sum()
    .to_frame('Non-Missing Count')
)

delay_analysis['Missing Count'] = (
    df[delay_reason_cols]
    .isna()
    .sum()
)

delay_analysis['Non-Missing %'] = (
    delay_analysis['Non-Missing Count']
    / len(df)
    * 100
).round(2)

display(delay_analysis)

,Non-Missing Count,Missing Count,Non-Missing %
CarrierDelay,98130,441617,18.18
WeatherDelay,98130,441617,18.18
NASDelay,98130,441617,18.18
SecurityDelay,98130,441617,18.18
LateAircraftDelay,98130,441617,18.18


In [35]:
delay_reason_relationship = pd.DataFrame({
    'Delayed Flights': [
        (df['ArrDelayMinutes'] > 0).sum()
    ],
    'Delay Reason Records': [
        df['CarrierDelay'].notna().sum()
    ],
    'Delayed Flights Without Reason': [
        (
            (df['ArrDelayMinutes'] > 0) &
            (df['CarrierDelay'].isna())
        ).sum()
    ],
    'Reason Records Without Arrival Delay': [
        (
            (df['CarrierDelay'].notna()) &
            (df['ArrDelayMinutes'] <= 0)
        ).sum()
    ]
})

display(delay_reason_relationship)

,Delayed Flights,Delay Reason Records,Delayed Flights Without Reason,Reason Records Without Arrival Delay
0,172133,98130,74003,0


In [36]:
delay_pattern = (
    df[delay_reason_cols]
    .notna()
    .value_counts()
    .reset_index(name='Flight Count')
)

display(delay_pattern)

,CarrierDelay,WeatherDelay,NASDelay,SecurityDelay,LateAircraftDelay,Flight Count
0,False,False,False,False,False,441617
1,True,True,True,True,True,98130


In [37]:
# DELAY ANALYSIS
delay_bucket_analysis = pd.DataFrame({
    'Category': [
        'No arrival delay',
        '1–14 min delay',
        '15–59 min delay',
        '60–119 min delay',
        '120+ min delay'
    ],
    'Flight Count': [
        (df['ArrDelayMinutes'] <= 0).sum(),
        ((df['ArrDelayMinutes'] > 0) & (df['ArrDelayMinutes'] < 15)).sum(),
        ((df['ArrDelayMinutes'] >= 15) & (df['ArrDelayMinutes'] < 60)).sum(),
        ((df['ArrDelayMinutes'] >= 60) & (df['ArrDelayMinutes'] < 120)).sum(),
        (df['ArrDelayMinutes'] >= 120).sum()
    ]
})

delay_bucket_analysis['% of Flights'] = (
    delay_bucket_analysis['Flight Count'] / len(df) * 100
).round(2)

display(delay_bucket_analysis)


,Category,Flight Count,% of Flights
0,No arrival delay,350136,64.87
1,1–14 min delay,74003,13.71
2,15–59 min delay,64885,12.02
3,60–119 min delay,19728,3.66
4,120+ min delay,13517,2.50


In [38]:
# REST OF COLUMNS 
# FirstDept
additional_time_cols=[
     'FirstDepTime',
    'TotalAddGTime',
    'LongestAddGTime'
]
additional_time_analysis = pd.DataFrame({
    'Column': additional_time_cols,
    'Non-Missing Count': [
        df[col].notna().sum()
        for col in additional_time_cols
    ],
    'Missing Count': [
        df[col].isna().sum()
        for col in additional_time_cols
    ],
    'Non-Missing %': [
        round(df[col].notna().mean() * 100, 2)
        for col in additional_time_cols
    ]
})

display(additional_time_analysis)

,Column,Non-Missing Count,Missing Count,Non-Missing %
0,FirstDepTime,4267,535480,0.79
1,TotalAddGTime,4267,535480,0.79
2,LongestAddGTime,4267,535480,0.79


In [39]:
additional_pattern = (
    df[additional_time_cols]
    .notna()
    .value_counts()
    .reset_index(name='Flight Count')
)

display(additional_pattern)


,FirstDepTime,TotalAddGTime,LongestAddGTime,Flight Count
0,False,False,False,535480
1,True,True,True,4267


In [40]:
additional_event_profile = (
    df.assign(
        HasAdditionalTime=df['FirstDepTime'].notna()
    )
    .groupby('HasAdditionalTime')['Cancelled']
    .agg(
        Total_Flights='size',
        Cancelled_Flights='sum'
    )
    .reset_index()
)

additional_event_profile['Cancellation %'] = (
    additional_event_profile['Cancelled_Flights']
    / additional_event_profile['Total_Flights']
    * 100
).round(2)

display(additional_event_profile)

,HasAdditionalTime,Total_Flights,Cancelled_Flights,Cancellation %
0,False,535480,15968,2.98
1,True,4267,344,8.06


In [41]:
# SINCE THEY ARE IMPORTANT FOR FUTURE WE ARE NOT DROPPING IT

In [42]:
remaining_high_missing = (
    missing_report[
        (missing_report['Missing %'] > 50) &
        (missing_report['Column'].isin(df.columns))
    ]
    [['Column', 'Missing Count', 'Missing %', 'Data Type']]
    .copy()
)

display(remaining_high_missing)

,Column,Missing Count,Missing %,Data Type
8,DivArrDelay,538849,99.83,float64
9,DivActualElapsedTime,538849,99.83,float64
10,Div1TailNum,538829,99.83,object
12,DivReachedDest,538581,99.78,float64
13,DivDistance,538581,99.78,float64
15,Div1Airport,538495,99.77,object
20,LongestAddGTime,535480,99.21,float64
21,TotalAddGTime,535480,99.21,float64
22,FirstDepTime,535480,99.21,float64
23,CancellationCode,523435,96.98,object


In [43]:
missing_report = (
    df.isna()
      .sum()
      .to_frame('Missing Count')
)

missing_report['Missing %'] = (
    df.isna().mean() * 100
).round(4)

missing_report['Data Type'] = df.dtypes

missing_report = (
    missing_report
    .reset_index()
    .rename(columns={'index': 'Column'})
    .sort_values('Missing %', ascending=False)
    .reset_index(drop=True)
)

display(missing_report.head(20))

,Column,Missing Count,Missing %,Data Type
0,DivArrDelay,538849,99.8336,float64
1,DivActualElapsedTime,538849,99.8336,float64
2,Div1TailNum,538829,99.8299,object
3,DivDistance,538581,99.7840,float64
4,DivReachedDest,538581,99.7840,float64
5,Div1Airport,538495,99.7680,object
6,LongestAddGTime,535480,99.2094,float64
7,FirstDepTime,535480,99.2094,float64
8,TotalAddGTime,535480,99.2094,float64
9,CancellationCode,523435,96.9778,object


In [44]:
# NOW FOR LOW MISSING VALUE COLUMNS
# ALL HAVE SAME EQAUL MISSING VALUE

In [45]:
arrival_missing_analysis = pd.DataFrame({
    'Group': [
        'Arrival data present',
        'Arrival data missing'
    ],
    'Flights': [
        df['ArrDelayMinutes'].notna().sum(),
        df['ArrDelayMinutes'].isna().sum()
    ],
    'Cancelled Flights': [
        df.loc[df['ArrDelayMinutes'].notna(), 'Cancelled'].sum(),
        df.loc[df['ArrDelayMinutes'].isna(), 'Cancelled'].sum()
    ]
})

arrival_missing_analysis['Cancellation %'] = (
    arrival_missing_analysis['Cancelled Flights']
    / arrival_missing_analysis['Flights']
    * 100
).round(2)

display(arrival_missing_analysis)

,Group,Flights,Cancelled Flights,Cancellation %
0,Arrival data present,522269,0,0.00
1,Arrival data missing,17478,16312,93.33


### INSIGHT
- total flights 17478 but cancelled only 16312 therefore 1166 remaining values must be related to diversions

In [46]:
# confirming
missing_arrival_profile = pd.DataFrame({
    'Metric': [
        'Missing arrival data',
        'Cancelled',
        'Not cancelled',
        'DivReachedDest populated',
        'DivDistance populated'
    ],
    'Count': [
        df['ArrDelayMinutes'].isna().sum(),
        (
            df['ArrDelayMinutes'].isna() &
            (df['Cancelled'] == 1)
        ).sum(),
        (
            df['ArrDelayMinutes'].isna() &
            (df['Cancelled'] == 0)
        ).sum(),
        (
            df['ArrDelayMinutes'].isna() &
            df['DivReachedDest'].notna()
        ).sum(),
        (
            df['ArrDelayMinutes'].isna() &
            df['DivDistance'].notna()
        ).sum()
    ]
})

display(missing_arrival_profile)

,Metric,Count
0,Missing arrival data,17478
1,Cancelled,16312
2,Not cancelled,1166
3,DivReachedDest populated,1166
4,DivDistance populated,1166


### INSIGHT
- FROM ABOVE MATRIX IT IS CONFIRMED THAT THE DATA GIVEN IS CORRECT BUT SOME VALUES ARE MISSING THUS THERE IS NO NEED TO DROP COLUMNS

# MEASURING DEPTTIME RELATED MEASURE

In [47]:
departure_missing_analysis = pd.DataFrame({
    'Group': [
        'Departure time present',
        'Departure time missing'
    ],
    'Flights': [
        df['DepTime'].notna().sum(),
        df['DepTime'].isna().sum()
    ],
    'Cancelled Flights': [
        df.loc[df['DepTime'].notna(), 'Cancelled'].sum(),
        df.loc[df['DepTime'].isna(), 'Cancelled'].sum()
    ]
})

departure_missing_analysis['Cancellation %'] = (
    departure_missing_analysis['Cancelled Flights']
    / departure_missing_analysis['Flights']
    * 100
).round(2)

display(departure_missing_analysis)

,Group,Flights,Cancelled Flights,Cancellation %
0,Departure time present,523861,426,0.08
1,Departure time missing,15886,15886,100.00


### INSIGHT
- Every flight with a missing departure time was cancelled.
- NOT EVERY cancelled FLIGHT was cancelled before departure

In [48]:
departure_cols = [
    'DepDelay',
    'DepDelayMinutes',
    'DepDel15',
    'DepartureDelayGroups',
    'WheelsOff',
    'TaxiOut'
]

departure_profile = pd.DataFrame({
    'Column': departure_cols,
    'Missing Count': [
        df[col].isna().sum()
        for col in departure_cols
    ],
    'Missing %': [
        round(df[col].isna().mean() * 100, 2)
        for col in departure_cols
    ],
    'Missing + DepTime Missing': [
        (
            df[col].isna() &
            df['DepTime'].isna()
        ).sum()
        for col in departure_cols
    ]
})

display(departure_profile)

,Column,Missing Count,Missing %,Missing + DepTime Missing
0,DepDelay,15923,2.95,15886
1,DepDelayMinutes,15923,2.95,15886
2,DepDel15,15923,2.95,15886
3,DepartureDelayGroups,15923,2.95,15886
4,WheelsOff,16227,3.01,15886
5,TaxiOut,16227,3.01,15886


### INSIGHT
- 15,886 records are explained by cancelled flights with no departure time.
- 37 additional missing → DepDelay block
- 341 additional missing → WheelsOff / TaxiOut

In [49]:
# CHECK FOR REST 37
departure_exceptions = pd.DataFrame({
    'Column': departure_cols,
    'Extra Missing Beyond DepTime': [
        (
            df[col].isna() &
            df['DepTime'].notna()
        ).sum()
        for col in departure_cols
    ],
    'Cancelled Among Exceptions': [
        (
            df[col].isna() &
            df['DepTime'].notna() &
            (df['Cancelled'] == 1)
        ).sum()
        for col in departure_cols
    ]
})

display(departure_exceptions)


,Column,Extra Missing Beyond DepTime,Cancelled Among Exceptions
0,DepDelay,37,37
1,DepDelayMinutes,37,37
2,DepDel15,37,37
3,DepartureDelayGroups,37,37
4,WheelsOff,341,341
5,TaxiOut,341,341


## BASIC UNDESTANDING
                        FLIGHT
                           │
             ┌─────────────┴─────────────┐
             │                           │
          CANCELLED                  OPERATED
             │                           │
       no departure                 departure
             │                           │
      DepTime = NaN               DepTime exists
             │                           │
      no arrival data             arrival / delay data

In [53]:
# FOR ARRIVAL RELATED MISSING VALUES
arrival_ops_cols = [
    'ArrTime',
    'WheelsOn',
    'TaxiIn'
]

arrival_ops_profile = pd.DataFrame({
    'Column': arrival_ops_cols,
    'Missing Count': [
        df[col].isna().sum()
        for col in arrival_ops_cols
    ],
    'Missing %': [
        round(df[col].isna().mean() * 100, 2)
        for col in arrival_ops_cols
    ],
    'Missing + Cancelled': [
        (
            df[col].isna() &
            (df['Cancelled'] == 1)
        ).sum()
        for col in arrival_ops_cols
    ],
    'Missing + Diverted': [
        (
            df[col].isna() &
            df['DivReachedDest'].notna()
        ).sum()
        for col in arrival_ops_cols
    ]
})

display(arrival_ops_profile)

,Column,Missing Count,Missing %,Missing + Cancelled,Missing + Diverted
0,ArrTime,16580,3.07,16312,268
1,WheelsOn,16580,3.07,16312,268
2,TaxiIn,16580,3.07,16312,268


In [54]:
arrival_performance_cols = [
    'ActualElapsedTime',
    'AirTime',
    'ArrDelay',
    'ArrDelayMinutes',
    'ArrDel15',
    'ArrivalDelayGroups'
]

arrival_performance_profile = pd.DataFrame({
    'Column': arrival_performance_cols,
    'Missing Count': [
        df[col].isna().sum()
        for col in arrival_performance_cols
    ],
    'Missing %': [
        round(df[col].isna().mean() * 100, 2)
        for col in arrival_performance_cols
    ]
})

display(arrival_performance_profile)

,Column,Missing Count,Missing %
0,ActualElapsedTime,17478,3.24
1,AirTime,17478,3.24
2,ArrDelay,17478,3.24
3,ArrDelayMinutes,17478,3.24
4,ArrDel15,17478,3.24
5,ArrivalDelayGroups,17478,3.24


# DATA AUDITING IS COMPLETED


#  DATA INTEGRITY
- DETECTING THINGS SUCH AS:-
- Negative flight duration
- Negative distance
- Negative taxi time
- Impossible delay values
- Arrival before departure
- Cancelled flight with impossible operational values

In [58]:
# FINDING NUMERIC COLS FIRST
numeric_cols=df.select_dtypes(include='number').columns.tolist()

print(f"Numeric columns: {len(numeric_cols)}")
display(pd.DataFrame({'Column': numeric_cols}))

Numeric columns: 54


,Column
0,Year
1,Quarter
2,Month
3,DayofMonth
4,DayOfWeek
5,DOT_ID_Reporting_Airline
6,Flight_Number_Reporting_Airline
7,OriginAirportID
8,OriginAirportSeqID
9,OriginCityMarketID


### FIRST INTEGRITY CHECK -- DISTANCE
IT CAN NEVER BE NEGATIVE

In [60]:
distance_check=pd.DataFrame({
    'Metrics':[
        'Missing',
        'Zero',
        'Positive',
        'Negative'  
    ],
    'Count':[
        df['Distance'].isna().sum(),
        (df['Distance']==0).sum(),
        (df['Distance']>0).sum(),
        (df['Distance']<0).sum()
    ]
})

display(distance_check)

,Metrics,Count
0,Missing,0
1,Zero,0
2,Positive,539747
3,Negative,0


### SECOND INTEGRITY CHECK -- FLIGHTS 
NUMBER TO FLIGHTS

In [65]:
flights_col=pd.DataFrame({
   'Metrics':  ['Missing',
    'Zero',
    'Positive',
    'Negative',
    'Unique'
            ],
    'Count':[
        df['Flights'].isna().sum(),
        (df['Flights']==0).sum(),
        (df['Flights']>0).sum(),
        (df['Flights']<0).sum(),
        (df['Flights'].nunique()),
    ]
}
    )
display(flights_col)

,Metrics,Count
0,Missing,0
1,Zero,0
2,Positive,539747
3,Negative,0
4,Unique,1


In [66]:
display(df['Flights'].value_counts().to_frame('Count'))

,Count
Flights,
1,539747


### INSIGHT
THE FLIGHTS COLUMN CONTAIN ONLY ONE VALUE FOR ENTIRE COLUMN THUS DROPPING IT

In [69]:
df=df.drop(columns=['Flights'])

In [70]:
df.shape

(539747, 70)

### SECOND INTEGRITY CHECK -- DISTANCEGROUP
CALCULATE DISTANCE

In [71]:
distance_group_check = pd.DataFrame({
    'Metric': [
        'Missing',
        'Zero',
        'Negative',
        'Unique Values',
        'Minimum',
        'Maximum'
    ],
    'Value': [
        df['DistanceGroup'].isna().sum(),
        (df['DistanceGroup'] == 0).sum(),
        (df['DistanceGroup'] < 0).sum(),
        df['DistanceGroup'].nunique(),
        df['DistanceGroup'].min(),
        df['DistanceGroup'].max()
    ]
})

display(distance_group_check)

,Metric,Value
0,Missing,0
1,Zero,0
2,Negative,0
3,Unique Values,11
4,Minimum,1
5,Maximum,11


In [72]:
### SINCE ALL VALUES ARE UNIQUE IT PASSES INTEGRITY TEST

### SECOND INTEGRITY CHECK -- TIME/CALENDAR FIELDS
- Month → 1–12
- DayofMonth → 1–31
- DayOfWeek → 1–7
- Quarter → 1–4

In [73]:
calendar_check = pd.DataFrame({
    'Column': [
        'Quarter',
        'Month',
        'DayofMonth',
        'DayOfWeek'
    ],
    'Min': [
        df['Quarter'].min(),
        df['Month'].min(),
        df['DayofMonth'].min(),
        df['DayOfWeek'].min()
    ],
    'Max': [
        df['Quarter'].max(),
        df['Month'].max(),
        df['DayofMonth'].max(),
        df['DayOfWeek'].max()
    ],
    'Missing': [
        df['Quarter'].isna().sum(),
        df['Month'].isna().sum(),
        df['DayofMonth'].isna().sum(),
        df['DayOfWeek'].isna().sum()
    ]
})

display(calendar_check)

,Column,Min,Max,Missing
0,Quarter,1,1,0
1,Month,1,1,0
2,DayofMonth,1,31,0
3,DayOfWeek,1,7,0


In [74]:
calendar_distribution = pd.DataFrame({
    'Month Count': df['Month'].value_counts().sort_index(),
    'Quarter Count': df['Quarter'].value_counts().sort_index()
})

display(calendar_distribution)

,Month Count,Quarter Count
1,539747,539747


# SINCE COLUMNS CONTAIN SINGULAR VALUE THUS DROPPING IT

In [75]:
constant_cols = ['Month', 'Quarter']

df = df.drop(columns=constant_cols)


In [76]:
df.shape

(539747, 68)

In [77]:
# YEAR CHECK
year_check = pd.DataFrame({
    'Metric': [
        'Unique Values',
        'Minimum',
        'Maximum',
        'Missing'
    ],
    'Value': [
        df['Year'].nunique(),
        df['Year'].min(),
        df['Year'].max(),
        df['Year'].isna().sum()
    ]
})

display(year_check)

display(df['Year'].value_counts().sort_index().to_frame('Flight Count'))


,Metric,Value
0,Unique Values,1
1,Minimum,2025
2,Maximum,2025
3,Missing,0


,Flight Count
Year,
2025,539747



### FROM YEAR COLUMN WE DETERMINE THIS DATASET BELONG TO JAN 2025

In [78]:
# VALIDATING IDENTIFIERS
flight_key = [
    'Year',
    'DayofMonth',
    'DayOfWeek',
    'Flight_Number_Reporting_Airline',
    'Origin',
    'Dest'
]

duplicate_flights = df.duplicated(
    subset=flight_key,
    keep=False
).sum()

print(f"Rows involved in duplicate flight keys: {duplicate_flights}")

Rows involved in duplicate flight keys: 398


### INSGIHT
FROM ABOVE CELL WE REACHED TO A CONCLUSION THAT THERE ARE 398 DUPLICATED ROWS WITH DUPLICATE IDS

In [79]:
duplicate_rows = df[
    df.duplicated(subset=flight_key, keep=False)
].sort_values(flight_key)

display(
    duplicate_rows[
        flight_key +
        ['CRSDepTime', 'DepTime', 'ArrTime', 'Cancelled', 'Diverted']
    ]
)

,Year,DayofMonth,DayOfWeek,Flight_Number_Reporting_Airline,Origin,Dest,CRSDepTime,DepTime,ArrTime,Cancelled,Diverted
75098,2025,1,3,15,JFK,SFO,1300,1253.0,1625.0,0,0
93252,2025,1,3,15,JFK,SFO,645,637.0,1028.0,0,0
1212,2025,1,3,305,MIA,SEA,1930,1934.0,2338.0,0,0
75238,2025,1,3,305,MIA,SEA,1735,1725.0,2135.0,0,0
75268,2025,1,3,347,MCO,SEA,1708,1710.0,2041.0,0,0
...,...,...,...,...,...,...,...,...,...,...,...
334182,2025,31,5,1861,AUS,PHX,1745,1740.0,1908.0,0,0
235166,2025,31,5,2042,ORD,MCO,1410,1410.0,1811.0,0,0
484696,2025,31,5,2042,ORD,MCO,2202,2150.0,145.0,0,0
235077,2025,31,5,2140,IAH,LGA,1625,1617.0,2036.0,0,0


## FROM ABOVE MATRIX
Year + DayofMonth + DayOfWeek + Flight Number + Origin + Destination is insufficient to uniquely identify a flight operation.

In [80]:
flight_key = [
    'Year',
    'DayofMonth',
    'Flight_Number_Reporting_Airline',
    'Origin',
    'Dest',
    'CRSDepTime'
]

duplicate_flights = df.duplicated(
    subset=flight_key,
    keep=False
).sum()

print(f"Rows involved in duplicate flight keys: {duplicate_flights}")

Rows involved in duplicate flight keys: 0


### Row grain: one scheduled flight operation identified by date, reporting-airline flight number, origin, destination, and scheduled departure time.

In [81]:
exact_duplicates = df.duplicated(keep=False).sum()

print(f"Exact duplicate rows: {exact_duplicates}")

Exact duplicate rows: 0


In [82]:
cancellation_profile = pd.DataFrame({
    'Metric': [
        'Total Cancelled',
        'Cancelled + DepTime',
        'Cancelled + ArrTime',
        'Cancelled + WheelsOff',
        'Cancelled + WheelsOn',
        'Cancelled + AirTime',
        'Cancelled + ArrDelay'
    ],
    'Count': [
        (df['Cancelled'] == 1).sum(),
        ((df['Cancelled'] == 1) & df['DepTime'].notna()).sum(),
        ((df['Cancelled'] == 1) & df['ArrTime'].notna()).sum(),
        ((df['Cancelled'] == 1) & df['WheelsOff'].notna()).sum(),
        ((df['Cancelled'] == 1) & df['WheelsOn'].notna()).sum(),
        ((df['Cancelled'] == 1) & df['AirTime'].notna()).sum(),
        ((df['Cancelled'] == 1) & df['ArrDelay'].notna()).sum()
    ]
})

display(cancellation_profile)

,Metric,Count
0,Total Cancelled,16312
1,Cancelled + DepTime,426
2,Cancelled + ArrTime,0
3,Cancelled + WheelsOff,85
4,Cancelled + WheelsOn,0
5,Cancelled + AirTime,0
6,Cancelled + ArrDelay,0


In [83]:
cancelled_departed_profile = pd.DataFrame({
    'Metric': [
        'Cancelled + DepTime',
        'Cancelled + WheelsOff',
        'Cancelled + TaxiOut',
        'Cancelled + AirTime',
        'Cancelled + WheelsOn',
        'Cancelled + ArrTime'
    ],
    'Count': [
        ((df['Cancelled'] == 1) & df['DepTime'].notna()).sum(),
        ((df['Cancelled'] == 1) & df['WheelsOff'].notna()).sum(),
        ((df['Cancelled'] == 1) & df['TaxiOut'].notna()).sum(),
        ((df['Cancelled'] == 1) & df['AirTime'].notna()).sum(),
        ((df['Cancelled'] == 1) & df['WheelsOn'].notna()).sum(),
        ((df['Cancelled'] == 1) & df['ArrTime'].notna()).sum()
    ]
})

display(cancelled_departed_profile)

,Metric,Count
0,Cancelled + DepTime,426
1,Cancelled + WheelsOff,85
2,Cancelled + TaxiOut,85
3,Cancelled + AirTime,0
4,Cancelled + WheelsOn,0
5,Cancelled + ArrTime,0


### CONCLUSION FROM CANCELLATION PROFILE
Cancelled: 16,312
     │
     ├── 15,886 → no DepTime
     │
     └── 426 → DepTime exists
                 │
                 ├── 341 → no WheelsOff
                 │
                 └── 85 → WheelsOff + TaxiOut
                              │
                              └── 0 → AirTime

In [84]:
# SIMILARLY FOR DIVERSION
diversion_profile = pd.DataFrame({
    'Metric': [
        'Total Diverted',
        'Diverted + Cancelled',
        'Diverted + DepTime',
        'Diverted + ArrTime',
        'Diverted + AirTime',
        'Diverted + DivReachedDest',
        'Diverted + DivDistance'
    ],
    'Count': [
        (df['Diverted'] == 1).sum(),
        ((df['Diverted'] == 1) & (df['Cancelled'] == 1)).sum(),
        ((df['Diverted'] == 1) & df['DepTime'].notna()).sum(),
        ((df['Diverted'] == 1) & df['ArrTime'].notna()).sum(),
        ((df['Diverted'] == 1) & df['AirTime'].notna()).sum(),
        ((df['Diverted'] == 1) & df['DivReachedDest'].notna()).sum(),
        ((df['Diverted'] == 1) & df['DivDistance'].notna()).sum()
    ]
})

display(diversion_profile)

,Metric,Count
0,Total Diverted,1166
1,Diverted + Cancelled,0
2,Diverted + DepTime,1166
3,Diverted + ArrTime,898
4,Diverted + AirTime,0
5,Diverted + DivReachedDest,1166
6,Diverted + DivDistance,1166


### DELAY RELATED METRICS 

In [86]:
delay_columns = [
    'DepDelay',
    'DepDelayMinutes',
    'ArrDelay',
    'ArrDelayMinutes'
]

delay_range_check = pd.DataFrame({
    'Column': delay_columns,
    'Missing': [
        df[col].isna().sum()
        for col in delay_columns
    ],
    'Negative': [
        (df[col] < 0).sum()
        for col in delay_columns
    ],
    'Zero': [
        (df[col] == 0).sum()
        for col in delay_columns
    ],
    'Positive': [
        (df[col] > 0).sum()
        for col in delay_columns
    ]
})

display(delay_range_check)

,Column,Missing,Negative,Zero,Positive
0,DepDelay,15923,325311,21405,177108
1,DepDelayMinutes,15923,0,346716,177108
2,ArrDelay,17478,341773,8363,172133
3,ArrDelayMinutes,17478,0,350136,172133


### COMPARISON BEWTEEN DEPARTURE AND ARRIVAL THAT CAUSES MOST FLIGHT DELAYS

In [91]:
delay_group_distribution = pd.DataFrame({
    'Departure': df['DepartureDelayGroups'].value_counts(dropna=False).sort_index(),
    'Arrival': df['ArrivalDelayGroups'].value_counts(dropna=False).sort_index()
})

display(delay_group_distribution)

,Departure,Arrival
-2.0,4513,162415
-1.0,320798,179358
0.0,102853,82366
1.0,33424,35325
2.0,17898,18477
3.0,11292,11083
4.0,7397,7535
5.0,5204,5256
6.0,3934,3949
7.0,2988,2988


In [92]:
departure_group_check = (
    df.groupby('DepartureDelayGroups', dropna=False)['DepDelayMinutes']
      .agg(['count', 'min', 'max'])
      .reset_index()
)

arrival_group_check = (
    df.groupby('ArrivalDelayGroups', dropna=False)['ArrDelayMinutes']
      .agg(['count', 'min', 'max'])
      .reset_index()
)

print("DEPARTURE DELAY GROUPS")
display(departure_group_check)

print("ARRIVAL DELAY GROUPS")
display(arrival_group_check)

DEPARTURE DELAY GROUPS


,DepartureDelayGroups,count,min,max
0,-2.0,4513,0.0,0.0
1,-1.0,320798,0.0,0.0
2,0.0,102853,0.0,14.0
3,1.0,33424,15.0,29.0
4,2.0,17898,30.0,44.0
5,3.0,11292,45.0,59.0
6,4.0,7397,60.0,74.0
7,5.0,5204,75.0,89.0
8,6.0,3934,90.0,104.0
9,7.0,2988,105.0,119.0


ARRIVAL DELAY GROUPS


,ArrivalDelayGroups,count,min,max
0,-2.0,162415,0.0,0.0
1,-1.0,179358,0.0,0.0
2,0.0,82366,0.0,14.0
3,1.0,35325,15.0,29.0
4,2.0,18477,30.0,44.0
5,3.0,11083,45.0,59.0
6,4.0,7535,60.0,74.0
7,5.0,5256,75.0,89.0
8,6.0,3949,90.0,104.0
9,7.0,2988,105.0,119.0


In [98]:
def expected_delay_group(delay):
    if pd.isna(delay):
        return np.nan
    elif delay <= -16:
        return -2
    elif delay < 0:
        return -1
    elif delay < 15:
        return 0
    elif delay >= 180:
        return 12
    else:
        return int((delay - 15) // 15) + 1


dep_expected = df['DepDelay'].apply(expected_delay_group)
arr_expected = df['ArrDelay'].apply(expected_delay_group)

validation = pd.DataFrame({
    'Column': [
        'DepartureDelayGroups',
        'ArrivalDelayGroups'
    ],
    'Mismatches': [
        (
            df['DepartureDelayGroups'].notna() &
            (df['DepartureDelayGroups'] != dep_expected)
        ).sum(),

        (
            df['ArrivalDelayGroups'].notna() &
            (df['ArrivalDelayGroups'] != arr_expected)
        ).sum()
    ]
})

display(validation)

,Column,Mismatches
0,DepartureDelayGroups,0
1,ArrivalDelayGroups,0


###  CHECKING FOR TIME FILEDS

In [99]:
time_cols = [
    'CRSDepTime',
    'DepTime',
    'WheelsOff',
    'WheelsOn',
    'CRSArrTime',
    'ArrTime'
]

time_range_check = pd.DataFrame({
    'Column': time_cols,
    'Missing': [df[col].isna().sum() for col in time_cols],
    'Min': [df[col].min() for col in time_cols],
    'Max': [df[col].max() for col in time_cols],
    'Unique': [df[col].nunique() for col in time_cols]
})

display(time_range_check)


,Column,Missing,Min,Max,Unique
0,CRSDepTime,0,5.0,2359.0,1209
1,DepTime,15886,1.0,2400.0,1396
2,WheelsOff,16227,1.0,2400.0,1400
3,WheelsOn,16580,1.0,2400.0,1440
4,CRSArrTime,0,1.0,2359.0,1310
5,ArrTime,16580,1.0,2400.0,1440


# THIS MATRIX NEEDS SOME VALIDATION AS MAX ARRTIME IS 2400 or 12 midnight

In [100]:
def invalid_hhmm(series):
    values = series.dropna()
    
    hours = values // 100
    minutes = values % 100
    
    return ((hours > 23) | (minutes > 59)).sum()


time_validation = pd.DataFrame({
    'Column': time_cols,
    'Invalid HHMM Count': [
        invalid_hhmm(df[col])
        for col in time_cols
    ]
})

display(time_validation)

,Column,Invalid HHMM Count
0,CRSDepTime,0
1,DepTime,33
2,WheelsOff,58
3,WheelsOn,202
4,CRSArrTime,0
5,ArrTime,245


In [101]:
# CHECK FOR INVALID
invalid_time_values = {}

for col in time_cols:
    values = df[col].dropna()
    
    invalid = values[
        ((values // 100) > 23) |
        ((values % 100) > 59)
    ]
    
    invalid_time_values[col] = invalid.value_counts().sort_index()

for col, counts in invalid_time_values.items():
    print(f"\n{col}")
    display(counts.to_frame('Count'))
    


CRSDepTime


,Count
CRSDepTime,



DepTime


,Count
DepTime,
2400.0,33



WheelsOff


,Count
WheelsOff,
2400.0,58



WheelsOn


,Count
WheelsOn,
2400.0,202



CRSArrTime


,Count
CRSArrTime,



ArrTime


,Count
ArrTime,
2400.0,245


### ALL INVALID HAS TIME 2400

In [102]:
time_2400_check = pd.DataFrame({
    'Column': time_cols,
    '2400 Count': [
        (df[col] == 2400).sum()
        for col in time_cols
    ]
})

display(time_2400_check)

,Column,2400 Count
0,CRSDepTime,0
1,DepTime,33
2,WheelsOff,58
3,WheelsOn,202
4,CRSArrTime,0
5,ArrTime,245


In [104]:
time_2400_rows = df[
    (df[time_cols] == 2400).any(axis=1)
][[
    'Year',
    b
    'DayofMonth',
    'Flight_Number_Reporting_Airline',
    'Origin',
    'Dest',
    'CRSDepTime',
    'DepTime',
    'WheelsOff',
    'WheelsOn',
    'CRSArrTime',
    'ArrTime',
    'Cancelled',
    'Diverted'
]].copy()

display(time_2400_rows)

,Year,DayofMonth,Flight_Number_Reporting_Airline,Origin,Dest,CRSDepTime,DepTime,WheelsOff,WheelsOn,CRSArrTime,ArrTime,Cancelled,Diverted
1139,2025,18,300,JFK,LAX,2100,2058.0,2123.0,2400.0,33,6.0,0,0
1877,2025,8,333,MIA,JAX,2255,2350.0,2400.0,51.0,23,55.0,0,0
1879,2025,10,333,MIA,JAX,2255,2400.0,13.0,106.0,23,116.0,0,0
1898,2025,29,333,MIA,JAX,2255,2245.0,2305.0,2400.0,23,4.0,0,0
2594,2025,9,376,MIA,ORD,2144,2145.0,2212.0,2400.0,27,8.0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...
538584,2025,7,847,PHL,LAX,2120,2136.0,2149.0,2400.0,23,24.0,0,0
538636,2025,19,865,AUS,FLL,2026,2027.0,2053.0,2400.0,5,22.0,0,0
538642,2025,30,865,AUS,FLL,2026,2023.0,2037.0,2353.0,5,2400.0,0,0
538752,2025,24,870,BOS,MCO,2029,2022.0,2040.0,2348.0,2359,2400.0,0,0


### THUS THIS SHOWS THAT 2400 IS ENTIRLY CORRECT

In [105]:
time_order_check = pd.DataFrame({
    'Check': [
        'DepTime > WheelsOff',
        'WheelsOff > WheelsOn',
        'WheelsOn > ArrTime'
    ],
    'Violations': [
        (
            df['DepTime'].notna() &
            df['WheelsOff'].notna() &
            (df['DepTime'] > df['WheelsOff'])
        ).sum(),

        (
            df['WheelsOff'].notna() &
            df['WheelsOn'].notna() &
            (df['WheelsOff'] > df['WheelsOn'])
        ).sum(),

        (
            df['WheelsOn'].notna() &
            df['ArrTime'].notna() &
            (df['WheelsOn'] > df['ArrTime'])
        ).sum()
    ]
})

display(time_order_check)

,Check,Violations
0,DepTime > WheelsOff,1396
1,WheelsOff > WheelsOn,22144
2,WheelsOn > ArrTime,1952


In [107]:
#  1hr =60 min conversion
def hhmm_to_minutes(series):
    hours = series // 100
    minutes = series % 100
    return hours * 60 + minutes


time_minutes = pd.DataFrame({
    'DepTime': hhmm_to_minutes(df['DepTime']),
    'WheelsOff': hhmm_to_minutes(df['WheelsOff']),
    'WheelsOn': hhmm_to_minutes(df['WheelsOn']),
    'ArrTime': hhmm_to_minutes(df['ArrTime'])
})

display(time_minutes.head())

,DepTime,WheelsOff,WheelsOn,ArrTime
0,416.0,439.0,604.0,613.0
1,412.0,442.0,615.0,622.0
2,412.0,425.0,592.0,603.0
3,413.0,437.0,606.0,616.0
4,415.0,441.0,608.0,613.0


In [108]:
# CONVERSION
chronological_times = time_minutes.copy()

# WheelsOff
chronological_times['WheelsOff'] = np.where(
    chronological_times['WheelsOff'] < chronological_times['DepTime'],
    chronological_times['WheelsOff'] + 1440,
    chronological_times['WheelsOff']
)

# WheelsOn
chronological_times['WheelsOn'] = np.where(
    chronological_times['WheelsOn'] < chronological_times['WheelsOff'],
    chronological_times['WheelsOn'] + 1440,
    chronological_times['WheelsOn']
)

# ArrTime
chronological_times['ArrTime'] = np.where(
    chronological_times['ArrTime'] < chronological_times['WheelsOn'],
    chronological_times['ArrTime'] + 1440,
    chronological_times['ArrTime']
)

display(chronological_times.head())

,DepTime,WheelsOff,WheelsOn,ArrTime
0,416.0,439.0,604.0,613.0
1,412.0,442.0,615.0,622.0
2,412.0,425.0,592.0,603.0
3,413.0,437.0,606.0,616.0
4,415.0,441.0,608.0,613.0


In [109]:
chronology_check = pd.DataFrame({
    'Check': [
        'DepTime <= WheelsOff',
        'WheelsOff <= WheelsOn',
        'WheelsOn <= ArrTime'
    ],
    'Violations': [
        (
            chronological_times['DepTime'].notna() &
            chronological_times['WheelsOff'].notna() &
            (chronological_times['DepTime'] > chronological_times['WheelsOff'])
        ).sum(),

        (
            chronological_times['WheelsOff'].notna() &
            chronological_times['WheelsOn'].notna() &
            (chronological_times['WheelsOff'] > chronological_times['WheelsOn'])
        ).sum(),

        (
            chronological_times['WheelsOn'].notna() &
            chronological_times['ArrTime'].notna() &
            (chronological_times['WheelsOn'] > chronological_times['ArrTime'])
        ).sum()
    ]
})

display(chronology_check)

,Check,Violations
0,DepTime <= WheelsOff,0
1,WheelsOff <= WheelsOn,4
2,WheelsOn <= ArrTime,6


In [111]:
## CHECKING INTEGRITY
exception_mask = (
    (
        chronological_times['WheelsOff'].notna() &
        chronological_times['WheelsOn'].notna() &
        (chronological_times['WheelsOff'] > chronological_times['WheelsOn'])
    )
    |
    (
        chronological_times['WheelsOn'].notna() &
        chronological_times['ArrTime'].notna() &
        (chronological_times['WheelsOn'] > chronological_times['ArrTime'])
    )
)

time_exceptions = df.loc[
    exception_mask,
    [
        'Year',
        
        'DayofMonth',
        'Flight_Number_Reporting_Airline',
        'Origin',
        'Dest',
        'DepTime',
        'WheelsOff',
        'WheelsOn',
        'ArrTime',
        'Cancelled',
        'Diverted',
        'CRSElapsedTime',
        'ActualElapsedTime',
        'AirTime',
        'TaxiOut',
        'TaxiIn'
    ]
].copy()

display(time_exceptions)

,Year,DayofMonth,Flight_Number_Reporting_Airline,Origin,Dest,DepTime,WheelsOff,WheelsOn,ArrTime,Cancelled,Diverted,CRSElapsedTime,ActualElapsedTime,AirTime,TaxiOut,TaxiIn
47391,2025,4,791,CLT,BNA,2348.0,2400.0,2358.0,6.0,0,0,96,78.0,58.0,12.0,8.0
191583,2025,23,4159,DTW,ATW,2348.0,7.0,2359.0,6.0,0,0,93,78.0,52.0,19.0,7.0
427325,2025,9,1662,SLC,LAS,2251.0,2324.0,2318.0,9.0,0,0,100,138.0,54.0,33.0,51.0
427326,2025,10,1662,SLC,LAS,2339.0,2354.0,2352.0,1.0,0,0,100,82.0,58.0,15.0,9.0
427328,2025,12,1662,SLC,LAS,2347.0,11.0,3.0,8.0,0,0,100,81.0,52.0,24.0,5.0
431469,2025,11,2034,ATL,BHM,2301.0,41.0,6.0,13.0,0,0,53,132.0,25.0,100.0,7.0
434829,2025,11,3097,ATL,VPS,2311.0,102.0,44.0,49.0,0,0,71,158.0,42.0,111.0,5.0
435570,2025,11,3121,ATL,BNA,2349.0,139.0,118.0,126.0,0,0,68,157.0,39.0,110.0,8.0
437500,2025,6,3184,ATL,JAN,2349.0,2400.0,2357.0,4.0,0,0,83,75.0,57.0,11.0,7.0
465433,2025,21,2501,ATL,MEM,2325.0,2342.0,2341.0,33.0,0,0,83,128.0,59.0,17.0,52.0


### CHECKING DURATION CALCULATIONS

In [112]:
duration_check = pd.DataFrame({
    'Metric': [
        'Rows with all duration components',
        'Exact duration matches',
        'Duration mismatches'
    ],
    'Count': [
        (
            df[['TaxiOut', 'AirTime', 'TaxiIn', 'ActualElapsedTime']]
            .notna()
            .all(axis=1)
        ).sum(),

        (
            (
                df['TaxiOut'] +
                df['AirTime'] +
                df['TaxiIn']
            ) == df['ActualElapsedTime']
        ).sum(),

        (
            (
                df['TaxiOut'] +
                df['AirTime'] +
                df['TaxiIn']
            ) != df['ActualElapsedTime']
        ).sum()
    ]
})

display(duration_check)

,Metric,Count
0,Rows with all duration components,522269
1,Exact duration matches,522269
2,Duration mismatches,17478


### INSIGHT
17,478 are exactly the flights with missing arrival data.

# IMPORTANT
### 17,478 flights lack duration data because arrival/operational information is unavailable, primarily due to cancellations/diversions. 

### The dataset contains 539,747 flights. Missing operational data is structurally associated with cancelled/diverted flights rather than arbitrary missingness. Duration components are exactly consistent for all 522,269 flights where they are available. Delay-group rules are consistent with their underlying delay-minute ranges. HHMM fields were validated, including midnight/2400 cases. Ten rare chronological inconsistencies remain and are retained as flagged records rather than artificially corrected.

In [113]:
df_analysis = df.copy()

# Delay indicators
df_analysis['IsDelayed'] = (
    df_analysis['ArrDelayMinutes'].fillna(0) > 0
).astype(int)

df_analysis['IsDelayed15'] = (
    df_analysis['ArrDel15'].fillna(0) == 1
).astype(int)

# Cancellation indicator
df_analysis['IsCancelled'] = (
    df_analysis['Cancelled'] == 1
).astype(int)

# Diversion indicator
df_analysis['IsDiverted'] = (
    df_analysis['Diverted'] == 1
).astype(int)

# Severe delay indicator
df_analysis['IsSeverelyDelayed'] = (
    df_analysis['ArrDelayMinutes'].fillna(0) >= 60
).astype(int)

print("Analysis dataset shape:", df_analysis.shape)

Analysis dataset shape: (539747, 73)


In [114]:
display(
    df_analysis[
        [
            'IsDelayed',
            'IsDelayed15',
            'IsCancelled',
            'IsDiverted',
            'IsSeverelyDelayed'
        ]
    ].sum().to_frame('Count')
)

,Count
IsDelayed,172133
IsDelayed15,98130
IsCancelled,16312
IsDiverted,1166
IsSeverelyDelayed,33245


## OVERALL SUMMARY

In [115]:
overall_summary = pd.DataFrame({
    'Metric': [
        'Total Flights',
        'Delayed Flights',
        'Delayed ≥15 min',
        'Severely Delayed ≥60 min',
        'Cancelled Flights',
        'Diverted Flights'
    ],
    'Count': [
        len(df_analysis),
        df_analysis['IsDelayed'].sum(),
        df_analysis['IsDelayed15'].sum(),
        df_analysis['IsSeverelyDelayed'].sum(),
        df_analysis['IsCancelled'].sum(),
        df_analysis['IsDiverted'].sum()
    ]
})

overall_summary['Percentage'] = (
    overall_summary['Count'] / len(df_analysis) * 100
).round(2)

display(overall_summary)

,Metric,Count,Percentage
0,Total Flights,539747,100.00
1,Delayed Flights,172133,31.89
2,Delayed ≥15 min,98130,18.18
3,Severely Delayed ≥60 min,33245,6.16
4,Cancelled Flights,16312,3.02
5,Diverted Flights,1166,0.22


In [ ]:
df_analysis.to_csv(
    'aviation_cleaned_JAN2025.csv',
    index=False
)
